app/demo-loop/Runner.cc

```diff
@@ -406,13 +406,10 @@ void Runner::build_diagnostics(RunnerInput const& inp)
     if (inp.action_diagnostic)
     {
         auto action_diagnostic = std::make_shared<ActionDiagnostic>(
-            core_params_->action_reg()->next_id(),
-            core_params_->action_reg(),
-            core_params_->particle(),
-            core_params_->max_streams());
+            core_params_->action_reg()->next_id());
 
         // Add to action registry
-        core_params_->action_reg()->insert(action_diagnostic);
+        core_params_->action_reg()->insert_mutable(action_diagnostic);
```

Comment:
I really dislike this, but `shared_ptr<ActionInterface const>` and `shared_ptr<ActionInterface>` become ambiguous overloads when passing `shared_ptr<DerivedAction>` or `shared_ptr<DerivedAction const>`.

Reply:
Do you dislike this more?
```c++
    // Register a mutable action
    template<class SPMut,
             std::enable_if_t<!std::is_const_v<typename SPMut::element_type>, bool>
             = true>
    void insert(SPMut action)
    {
        CELER_EXPECT(action);
        if (dynamic_cast<BeginRunActionInterface*>(action.get()))
        {
            mutable_actions_.push_back(action);
        }
        return this->insert_impl(std::move(action));
    }

    // Register a const action
    template<class SPConst,
             std::enable_if_t<std::is_const_v<typename SPConst::element_type>, bool>
             = true>
    void insert(SPConst action)
    {
        CELER_EXPECT(action);
        CELER_VALIDATE(
            !dynamic_cast<BeginRunActionInterface const*>(action.get()),
            << "begin-run action '" << action->label()
            << "' (ID=" << action->action_id().unchecked_get()
            << ") cannot be registered as const");
        return this->insert_impl(std::move(action));
    }
```

Reply:
Actually that's great! I think we can make it more readable with a helper traits class. Thanks!

Reply:
Concepts would probably make it even more readable whenever we can use c++ 20.